### ISSUE TITLE
negate ignores

### ISSUE THREAD

**Issue by chpio** [babel/babel#5086]
### Babel Configuration (.babelrc, package.json, cli command)
```json
{
  "presets": ["es2015"],
	"ignore": [
		"node_modues/**",
		"!node_modules/foobar/src/**"
	]
}
```

### Expected Behavior
I expect babel to handle the files in `node_modules/foobar/src/**`.

### Current Behavior
Babel isn't even handling files *not* included in `node_modules`.

### Possible Solution
Use another "globbing" plugin which supports negation instead of *minimatch*.

* https://github.com/gulpjs/glob-stream/blob/bb047a0e378e8ba838231e8a1c19f9e9c27c480e/index.js#L46-L61
* https://github.com/gulpjs/glob-stream/blob/bb047a0e378e8ba838231e8a1c19f9e9c27c480e/index.js#L98-L108

### Context
Compile an ES6-module inside of `node_modules`.

### Your Environment

| software         | version
| ---------------- | -------
| Babel            |  6.18.0 |
| node             | 7.4.0 |
| npm              | 4.0.5 |
| Operating System | Linux |

**Comment by xtuc** [comment]
To be clear, assuming the following directory structure:

```
node_modules
├── fizz
└── foobar
    └── src
```

You want to ignore `fizz` (and the other modules) but transpile files in `foobar`.
In Babel we don't have a way to exclude one directory but still include specific subdirectories.

**Comment by traviskaufman** [comment]
I'd be interested in working on this feature! 

Would the team be opposed to using a module such as [parse-gitignore](https://www.npmjs.com/package/parse-gitignore) to handle ignore patterns? This would resolve the current issue, as well as allow `"ignore"` to accommodate any type of gitignore-style pattern, which IMO is more expressive than glob-star patterns. 

I could also see why that may be considered _too_ heavy-handed of an approach. In that case, there's a module called [extglob](https://www.npmjs.com/package/extglob) which may be able to accommodate this case.

**Comment by xtuc** [comment]
Great, thanks.

We need to ensure backward compatibility in config files. If the API is the same, you can use it.

**Comment by traviskaufman** [comment]
According to git's [pattern format rules](https://git-scm.com/docs/gitignore#_pattern_format), the API seems to mirror glob/fnmatch's API, so it should be okay. If you see specific instances where gitignore matching might be incompatible with globstar matching, I can write test cases for it to ensure it doesn't break BC.

**Comment by xtuc** [comment]
Ok, this looks great.

We also need to keep the documentation up to date. I can do it if you don't want to.

**Comment by traviskaufman** [comment]
So it seems like the root of this issue comes from the fact that files determine whether or not they should be ignored by iterating through a list of regexes and if any one regex matches (or none match, in the case of `only`), it excludes itself. 

This is problematic in terms of accomplishing the goals of this issues since regexes can't be independent of one another in order to implement more sophisticated ignore patterns. For example: an `ignore` option with a value of `["node_modules/**", "!node_modules/foobar/src/**"]` would be transformed into two _distinct_ regex via `minimatch.makeRe`, and thus any file under `node_modules/foobar/src/` would still be flagged as one to be excluded due to the first regex. Note that due to [how this code is implemented](https://github.com/babel/babel/blob/master/packages/babel-core/src/util.js#L119-L128), ordering of regexes would not make a difference.

The second issue with regexes is that some glob patterns - especially dependent glob patterns such as ones specified above - could be unable to be expressed via a regex. 

So my recommendation would be to move to a more sophisticated mechanism for ignores rather than regex. It looks like the ignore logic is [already set up to accept a function](https://github.com/babel/babel/blob/master/packages/babel-core/src/util.js#L138), so instead of having an array of regexes, we could some sort of `FileFilter` class or utility that can handle globs, regexes, etc. I like this approach because not only is it robust but it's an internal change so we can still maintain backwards compatibility. 

An alternative solution would be to detect all of the ignores that are globs, and try and use some mechanism for determining an appropriate regex that takes dependent glob patterns into account. 

Thoughts?

**Comment by xtuc** [comment]
Yes, good analyze :+1: 

Your recommendation makes sense.

**Comment by traviskaufman** [comment]
Haven't forgotten about this, just been v. busy :) Hopefully will have time this weekend to take it on!

**Comment by hzoo** [comment]
This should be fixed in Babel 7 with https://github.com/babel/babel/pull/5487 - we are using https://github.com/jonschlinkert/micromatch which seems to accept negation?

can test with: 
"babel-core": "7.0.0-alpha.3"
"babel-cli": "7.0.0-alpha.3"
"babel-register": "7.0.0-alpha.5"

**Comment by loganfsmyth** [comment]
I actually didn't cover that in the current PR. Still could be useful though.

**Comment by rajatvijay** [comment]
Hi everyone

I would like to work on this issue. 

> This should be fixed in Babel 7 with #5487 - we are using https://github.com/jonschlinkert/micromatch which seems to accept negation?

Is it a good idea to use it (https://github.com/jonschlinkert/micromatch) in Babel 6?

**Comment by loganfsmyth** [comment]
Woops, missed this comment, sorry. Just put up this PR: https://github.com/babel/babel/pull/5625
